WordPress/WordPressTest/TestContextManager.m

```diff
@@ -6,19 +6,28 @@
 // Based on the Feature Flag value
 @implementation TestContextManager
 
-static TestContextManager *_instance;
-
 - (instancetype)init
 {
     self = [super init];
     if (self) {
-        // Override the shared ContextManager
-        _stack = [[ContextManagerMock alloc] init];
+        self.stack = [[ContextManagerMock alloc] init];
     }
 
     return self;
 }
 
+- (void)setStack:(id<ManagerMock, CoreDataStack>)stack
+{
+    if (stack == _stack) {
+        return;
+    }
+
+    _stack = stack;
+    // Override the shared ContextManager
+    [ContextManager internalSharedInstance];
```

Comment:
The short answer is, I kept the [original code](https://github.com/wordpress-mobile/WordPress-iOS/pull/18512/files#diff-7b102973d3b2dcb347034552ed9c073ba74e2a870795fc06ad671ec905fd6ce9) in `ContextManagerMock`.

As for the longer version... It's a bit complicated, I'll try to explain...

------

To understand why this is needed, we'll need to go back to the dispatch_once calls in [`ContextManager.mainContext`](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L75) and [`writerContext`](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L63).

The issue they have caused is, only _the very first instance_ that calls `mainContext` or `writerContext` gets a valid value, all other instances get nil. Bear in mind, the "instance" here also includes instances of `ContextManager` subclass `ContextManagerMock`.

Let's discuss a hypothetical situation first. Say there are two such instances created at runtime: fakeManagerA and realManagerB. And following calls are made, in their listed order:
1. `fakeManagerA.writerContext`
    As a result, fakeManagerA gets a valid `writerContext` value.
1. `realManagerB.writerContext`.
    As a result, realManagerB gets a nil `writerContext`, since fakeManagerA called `writerContext` first.
1. `realManagerB.mainContext`.
    BOOM. Crash. Due to exception "Parent NSManagedObjectContext must not be nil".
    Why? Because `mainContext` [uses `writerContext` as its parent context](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L77), which is nil as mentioned above.

Of course, this isn't a real issue in the app, for two facts:
1. There is only one single instance of `ContextManager`.
1. `writerContext` is a private API, no way the first call to it can be made by anyone else other than `mainContext`.

Unfortunately both are not facts in unit test.
1. Test cases can create `TestContextManager` whenever they want, which means there are [many instances of `ContextManagerMock`](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/WordPressTest/TestContextManager.m#L16), subclass of `ContextManager`.
2. `ContextManagerMock` [overrides `mainContext`](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/WordPressTest/ContextManagerMock.m#L67-L77), which means _`ContextManager.mainContext`_ method is _not called_ when the instance is a `ContextManagerMock`, so it's very much possible that `ContextManager.writerContext` is called _without calling_ `ContextManager.mainContext`, like by [this notification handler](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L258-L265).

Now, let's go back to the hypothetical situation discussed above. Consider fakeManagerA as an instance of `ContextManagerMock`, realManagerB as the `ContextManager` singleton. And following calls are made, in their listed order:
1. `fakeManagerA.mainContext` is called.
    As mentioned above, this method returns the context created by `ContextManagerMock.mainContext` override, not `ContextManager.mainContext`.
1. `fakeManagerA.mainContext.save()` is called.
    As a side effect of this, [the notification hander](https://github.com/wordpress-mobile/WordPress-iOS/blob/444a8fe10ad01d4aed4ea2a1d650c4062ec56aba/WordPress/Classes/Utility/ContextManager.m#L258-L265) is now called by CoreData framework.
1. `fakeManagerA.writerContext` is called by above notification handler.
    As a result, fakeManagerA gets a valid `writerContext` value.
1. (Looks familiar? Yes, you have seen this "~hypothetical~ situation" before)
1. `realManagerB.writerContext` is called.
    As a result, realManagerB gets a nil value, since fakeManagerA calls writerContext first.
1. `realManagerB.mainContext` is called. BOOM. Unit test crashes, for reason discussed before.

A workaround for this is, making sure realManagerB, the `ContextManager` singleton, calls `mainContext` and `writerContext` first. Which is why you see this statement `[ContextManager internalSharedInstance]` here, to force creating `ContextManager` singleton[^1].

Apologise this turns into an essay... But I appreciate you reading this far, what an adventure!

[^1]: As I mentioned in P2, both `mainContext` and `writerContext` are created during initialisation, which means their `dispatch_once` calls are redundant.

Reply:
> Apologise this turned into an essay... But I appreciate you reading this far, what an adventure!

It was a fun read. Thank you 😄